### PR TITLE
Node Reaper: Ghost/Unjoined Nodes

### LINKED ISSUE #12: Feature: Terminate EC2 instances that never join the cluster
I've observed this several times, a node fails to join the cluster for some reason (e.g. got ec2 on bad physical host, etc), however the EC2 is still hanging around, this is bad because:
- Autoscaling Group considers it a healthy node
- Waste of $$

We should consider:
- scan all cluster scaling groups
- scan all instances belonging to scaling groups
- if matching node is not found AND ec2 instance has been running for more than N minutes, terminate the ec2 instance

### PR DESCRIPTION
**Pull request by eytan-avisror**
Fixes #12 #11 

This handles unjoined & ghost nodes.

## Testing
Did manual testing to reproduce scenarios, the scenario for ghost nodes is hard to reproduce so was only able to unit test it.

### CODE REVIEW

File: pkg/reaper/nodereaper/nodereaper.go [hunk 1]
```diff
@@ -101,8 +107,33 @@ func (ctx *ReaperContext) validateArguments(args *Args) error {
 		return err
 	}
 	ctx.TimeToReap = args.ReapAfter
+
+	if args.ReapUnjoined {
+		if args.ReapUnjoinedThresholdMinutes < 10 {
+			err := fmt.Errorf("--reap-after must be set to a number greater than or equal to 10")
```
[the remark this answers is not in the document]
Comment (eytan-avisror) [reply pkg/reaper/nodereaper/nodereaper.go]: good catch :)

**Comment by davemasselink** [comment]
hmm... could have sworn I was a part of the authorized-approvers. not sure why it's not counting me there

**Comment by eytan-avisror** [comment]
@davemasselink I am sure I added you.. not sure what happened to your access - will re-add

File: pkg/reaper/nodereaper/nodereaper.go [hunk 2]
```diff
@@ -101,8 +107,33 @@ func (ctx *ReaperContext) validateArguments(args *Args) error {
 		return err
 	}
 	ctx.TimeToReap = args.ReapAfter
+
+	if args.ReapUnjoined {
+		if args.ReapUnjoinedThresholdMinutes < 10 {
+			err := fmt.Errorf("--reap-unjoined-threshold-minutes must be set to a number greater than or equal to 10")
```
Comment (shrinandj) [inline pkg/reaper/nodereaper/nodereaper.go]: Nice!

File: pkg/reaper/nodereaper/nodereaper_test.go [hunk 3]
```diff
@@ -1114,6 +1194,182 @@ func TestMaxKill(t *testing.T) {
 	testCase.Run(t, false)
 }
 
+func TestUnjoinedPositive(t *testing.T) {
```
Comment (shrinandj) [inline pkg/reaper/nodereaper/nodereaper_test.go]: As always, the tests look incredible! 💯

File: pkg/reaper/nodereaper/nodereaper.go (hunk 2 above)
Comment (eytan-avisror) [reply pkg/reaper/nodereaper/nodereaper.go]: better safe than sorry

File: pkg/reaper/nodereaper/nodereaper.go [hunk 4]
```diff
@@ -289,8 +327,85 @@ func (ctx *ReaperContext) deriveFlappyDrainReapableNodes() error {
 	return nil
 }
 
+// Handle ghost-reapable nodes
+func (ctx *ReaperContext) deriveGhostDrainReapableNodes(w ReaperAwsAuth) error {
+	log.Infoln("scanning for ghost drain-reapable nodes")
+	for instance, node := range ctx.NodeInstanceIDs {
```
Comment (shrinandj) [inline pkg/reaper/nodereaper/nodereaper.go]: This is going to make 2 AWS API calls for every node in the cluster. On large/busy clusters, the AWS API calls might start getting throttled. Is there a way to optimize this?

Make 2 AWS API calls to get ALL "Terminated instances" in the cluster (with the KubernetesCluster tag) and check the output against all nodes in the cluster. Would this work?

Comment (eytan-avisror) [reply pkg/reaper/nodereaper/nodereaper.go]: You're right, I can probably make a single Describe call before the loop, and do implement the filtering in governor.
It's a bit more code, but for good cause

Comment (eytan-avisror) [reply pkg/reaper/nodereaper/nodereaper.go]: Fixed by caching all instances and implementing filtering

**Pull request merged**